source/adminguide/extensions/developer.rst

```diff
@@ -120,7 +120,53 @@ Action Lifecycle
 1. A CloudStack action (e.g., deploy VM) triggers a corresponding extension action.
 2. CloudStack invokes the extension’s executable with appropriate parameters.
 3. The extension processes the input and responds within the timeout.
-4. CloudStack continues orchestration based on the result.
+4. CloudStack continues action workflow based on the result.
+
+Console Access for Instances with Orchestrator Extensions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Orchestrator extensions can provide console access for instances either through **VNC** or a **direct URL**.
+To enable this, the extension must implement the ``getconsole`` action and return output in one of the following JSON formats:
+
+VNC-based console:
+
+.. code-block:: json
+
+    {
+      "status": "success",
+      ...
+      "console": {
+        "host": "pve-node1.internal",
+        "port": "5901",
+        "password": "PVEVNC:6329C6AA::ZPcs5MT....d9",
+        "passwordonetimeuseonly": true
+        "protocol": "vnc"
+      }
+    }
+
+``passwordonetimeuseonly`` is optional. It can be set to ``true`` if the system returns a one-time-use VNC ticket.
+
+For VNC-based access, the returned details are forwarded to the Console Proxy VM (CPVM) in the same zone as the instance. The specified **host** and **port** must be reachable from the CPVM.  
+
+Direct URL-based console:
+
+.. code-block:: json
+
+    {
+      "status": "success",
+      ...
+      "console": {
+        "url": "CONSOLE_URL",
+        "protocol": "direct"
+      }
+    }
+
+
+.. note::
+   For direct URL–based console access, CloudStack does not report the acquired or client IP address.
+   In this mode, security and access control must be handled by the server providing the console.
+
+   Protocol value for direct URL–based console access can be any of the following value - link, direct, url.
```

Comment:
What's the difference between the three direct URL based console access? How is link, direct, url different ?